### PR TITLE
OSD-GDB server: Provides interface to connect it with GDB

### PR DESCRIPTION
**Pull request by shivmgg**
The current program provides the API to support following features:
(1) Connect to GDB over TCP
(2) Send and Recieve data to/from the client
(3) Receive RSP packet from the client
(4) Validate the obtained packet using checksum
(5) Send RSP packet to the client

I will update the code with other essential features in the meantime.

### CODE REVIEW

**Comment by imphil** [comment]
Thanks for this PR! Before we go into the details, let's start with some fundamental things.

- Please separate between public and private API. The public API goes into the .h file and has corresponding functions without "static" in the .c file, but marked with the API_EXPORT macro. On the other hand private functions are only in the .c file and marked with "static".
  This is especially important to quickly see the entry points to your class.

- Even though the class is named "osd_gdbserver" the files are typically named "gdbserver". Please look at the existing code for these naming conventions.

- Please take e.g. the osd_systracelogger class as an example how to make a class out of your collection of functions. This includes the _new() and _free() functions and the context object (i.e. the "this" pointer). This should allow you to get rid of the "connection" struct. All functions the public API should be prefixed with "osd_gdbserver_", the private ones are typically not.

- The naming of the functions could be a bit more intuitive. "get" and "put" are typically operations on a list or similar data structures, not on a network socket. A "packet" is ambiguous in OSD since we also have the debug packets around. 

- We need to work out how to handle blocking, i.e. where blocking is expected, and where the API should/must be asynchronous. I guess this will become clearer once we have a first application actually using the class. Which brings us to the next point.

- Testing: Unit tests are required. You can start by testing the functions which don't require a network socket, but at some point in time you'll need to create a "mock gdb client", i.e. code which actually connects to your TCP port. You can then instruct this code to send specific requests and responses. Examples of such functionality are in the source code (e.g. mock_hostmod).

- Development is much easier if you start right away creating an application using your new class. If you look at osd-systrace-log for example you see that it's not hard (most of the code is boilerplate and can be copied).



So that's a couple points that I'd like to see worked on first. Once this is done we can focus on the smaller things, and talk about how to actually implement the handling of commands coming in from GDB.

**Comment by shivmgg** [comment]
@imphil Thanks a lot for the review.
I have tried my best to cover everything as stated in the first few points. 

> Testing: Unit tests are required. You can start by testing the functions which don't require a network socket, but at some point in time you'll need to create a "mock gdb client", i.e. code which actually connects to your TCP port. You can then instruct this code to send specific requests and responses. Examples of such functionality are in the source code (e.g. mock_hostmod).

I will add unit tests for send/receive_rsp_packet() functions. The idea of writing a "mock gdb client" is a little frightening but still exciting right now. I will look into the code of mock_hostmod() for more details.

> Development is much easier if you start right away creating an application using your new class. If you look at osd-systrace-log for example you see that it's not hard (most of the code is boilerplate and can be copied).

I think you meant "osd-target-run". Please correct me, if I am wrong. I will start with it right away.

**Comment by imphil** [comment]
The challenges you have in testing are indeed tricky and require a bit of thought. Designing software in a way that it is testable isn't always easy, but will lead to a much clearer software design in the end.

In general, unit tests test the public API of a class. For us that's all functions which appear in the include/osd/*.h files, which are not marked "static" and have the API_EXPORT macro set (which makes the functions visible outside the library). 

For our osd_gdbserver class the public API is not exposing the functions which parse and assemble the stream of GDB commands. But in fact this code is the most tricky to write and benefits most from unit tests. So we should find a way to make these functions testable.

So what options do we have? The easiest way is to make the functions we want to test "public", at least when running the unit tests. I'd hence go ahead and do the following:
- Don't mark the parsing and assembling functions static.
- Create a header file (e.g. src/gdbserver-private.h) in which you declare the functions. (You don't need to add this header file to a Makefile.)
- Include this private header file in the unit test: "#include "../../src/libosd/gdbserver-private.h"`

You can now call such "private" functions in your test code. 

Prerequisite, however, is you don't have a [Matryoshka doll](https://en.wikipedia.org/wiki/Matryoshka_doll)-like design in your functions, i.e. functions which call other subfunctions in a nesting fashion. An example is the function `send_rsp_packet()`. Right now it has two functions: assemble the data to send according to the RSP protocol, and send it over the TCP connection. To test this function you need to either have the full TCP connection tested with it, or somehow intercept the sending functions, both not good options. Things get easier if you create two functions instead. One assembles the output string according to the RSP protocol, and returns the buffer. And the other one then sends this buffer over TCP. This makes it much easier to test the first function, which is the more tricky one.

**Comment by shivmgg** [comment]
Thanks a lot for the guidance. 

> Prerequisite, however, is you don't have a Matryoshka doll-like design in your functions, i.e. functions which call other subfunctions in a nesting fashion. An example is the function send_rsp_packet(). Right now it has two functions: assemble the data to send according to the RSP protocol, and send it over the TCP connection. To test this function you need to either have the full TCP connection tested with it, or somehow intercept the sending functions, both not good options. Things get easier if you create two functions instead. One assembles the output string according to the RSP protocol, and returns the buffer. And the other one then sends this buffer over TCP. This makes it much easier to test the first function, which is the more tricky one.

Some more issues:
- Almost every function needs access to "osd_gdbserver_ctx" struct object. The object's member elements like "buffer", etc. are initialised using osd_gdbserver_read_data() function. This function cannot be invoked right now. 
- One alternative solution could be to have a set_osd_gdbserver_ctx() function and set these values manually during testing. Similarly, having a get_osd_gdbserver_ctx() function might help.

**Comment by imphil** [comment]
Your version in the edit got it right. Functions like send_rsp_packet_inner() (better name needed) don't need the context struct and hence can easily be tested on its own. Please go ahead and try to update the PR in this style (and add the tests).

**Comment by imphil** [comment]
I forgot to mention: you don't need to make all functions which are currently static functions, i.e. helpers, available for testing. But you absolutely should test the functions which implement the RSP protocol and deal with string buffers, since those functions are very easy to get wrong (functional errors, buffer overflows, other memory errors, etc.)

**Comment by shivmgg** [comment]
@imphil Apologies for the delay. 
I have added unit tests for the private functions. Please take a look whenever you get some time.
I think I should start working on the implementation of different RSP commands.

**Comment by shivmgg** [comment]
Thanks for the review. I have incorporated almost every change requested. I have also added two more unit tests to check some corner cases. Here's a brief description of the two functions:

- **validate_rsp_packet()**: This function traverses through RSP packet "$packet-data#checksum" starting from '$' till '#'. It then calculates the checksum of the obtained packet-data and compares it with the received checksum. This function ensures that the packet-data obtained is valid and uncorrupted.
- **encode_rsp_packet()** (name changed): This function is responsible for packaging of the packet-data into RSP packet format. It also calculates the checksum of the packet data to be sent to the client (GDB).

Regarding the size of **buffer**, we can keep it constant and equal to the MAX_GDB_PACKET_SIZE, i.e., 1024 (currently). This is the maximum size of the packet received/send from/to the GDB via TCP. Also, the commands we are supporting right now will suffice with this size.

**Comment by shivmgg** [comment]
I have added code implementation of following GDB commands:
-  'g'            : read general registers
-  'G XX'      : write general registers
-  'p n '        : read the value of a specific register n
-  'P n..=r..'  : write a specific register n with value r

Next, I will add support for memory read/write command and then, connect GDB using target command. I need to complete the following RSP packet exchanges for the GDB target remote command to support connectivity. 

image

**Comment by shivmgg** [comment]
@imphil  Can you please suggest some way to connect to the actual GDB (client)? I have added support for almost every other command required to complete the RSP packet exchanges for the GDB **target** command. Is it possible using unit tests?

**Comment by imphil** [comment]
Unit testing can and needs to be done through a mock gdb client. I already outlined that in one of the previous discussions, please implement that. Before such testing is in place I will not have a further look at the code.

**Comment by shivmgg** [comment]
Thanks a lot for the response. 
Regarding the implementation of mock GDB client, it should be similar to mock_hostmod.c and should have features of the actual GDB, right?

**Comment by imphil** [comment]
It should open a tcp connection, and act like a gdb client, by sending commands that you told the mock client to do, and verifying the response.

**Comment by shivmgg** [comment]
I have added a rough implementation of mock GDBclient program and unit tests to check and validate the TCP connection between GDB server and mock GDBclient program. The test cases are working fine. Currently, I used PThread API to establish a TCP connection. Next, I am going to work on the non-blocking aspect of the program using the **Event Poll interface.** 

The non-blocking TCP server can be implemented as follows:
1. Create a socket 
2. Mark it as non-blocking
3. Bind it to an address
4. Listen on the socket
5. Create an epoll instance
6. Add socket to the epoll instance (this way the incoming requests can be monitored through event notification)
7. Create a read event queue
8. Create threads for processing tasks from read queue
9. Create a write event queue
10. Create threads for processing tasks in the write queue
11. Wait for events on epoll instance in a loop
12. For incoming requests events
            -  call accept
            -  mark the accepted socket as non-blocking
            -  add it to the epoll instance for monitoring
13. For read events, push the file descriptor and user data to read event queue
14. For write events, push the file descriptor and user data to write event queue
15. For close events, remove the file descriptor from the epoll instance.

I think the above execution is similar to what is implemented in mock_hostcontroller program. 
Please check if the above execution will work or not.

File: src/libosd/include/osd/gdbserver.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ */
+#define OSD_GDBSERVER_PORT 5555
+
+/**
+ * Indicates the size of the buffer
+ */
+#define OSD_GDBSERVER_BUFF_SIZE 1024
+
+/**
+ * Create a new context object
```
[the remark this answers is not in the document]
Comment (shivmgg) [reply src/libosd/include/osd/gdbserver.h]: Resolved.

**Comment by shivmgg** [comment]
I have tried to resolve some major issues in this iteration.  

> Path 1: gdb-client initiated commands:
GDB client -> osd_gdbserver -> (some function handling the command, typically using osd_hostmod) -> ... (response)
Questions to be answered:
> -  Which things can happen in parallel (i.e. are asynchronus)
> -  Is the gdbserver protocol a synchronous (blocking) or asynchronous protocol?
>    Example: if the gdb client sends a command to read memory from the target, is the answer sent when the memory has been read, or does it return immediately (e.g. with an acknowledgement saying "command receive"), and at some later point a new message is sent with the data?
> -  Answering the previous question is essential to the design of the osd_gdbserver module: it determines if you need to (potentially) handle multiple commands in parallel, or just one. 

Generally speaking, most commands in the OSD-GDB server are sequential in nature. Each packet received from the GDB (client) should be acknowledged with a single character.
-  '+' to indicate satisfactory receipt, with the valid checksum.
-  '-' to indicate failure and request retransmission. 

For instance, the following image illustrates the sequence diagram of "read memory / m packet" request. 

final-page-2

As per my understanding, we don't need to handle multiple commands simultaneously. 

> Path 2: event notifications (e.g. breakpoint notifications)
osd_hostmod (events) -> osd_gdbserver -> (TCP send) -> GDB client

> osd_hostmod provide a polling interface (osd_hostmod_event_receive()) and a asynchronous interface (register a event rx callback with osd_hostmod_new()). Depending on the answer of the remaining design both could be used.
Questions to be answered:
> - How are these notifications being combined with the gdb client-initiated data? I.e. breakpoint notifications be sent to a gdb client at any time, or only when no other command (see above) is being processed? 

The GDB command to set breakpoints, break does not immediately cause a RSP interaction. GDB only actually sets breakpoints immediately before execution (for example by a continue or step command) and immediately clears them when a breakpoint is hit. Event notifications, i.e. in case of breakpoint hit are sent to the GDB only when no other command is being processed. This is again a blocking/synchronous process. 

-  I think the challenge here is to see how hostmod and GDB client will interact with osd_gdbserver. Every process/transaction in the server is blocking/synchronous and hence, the osd_gdbserver needs to switch from one program to another at any point in time instead of working on them simultaneously.

**Comment by imphil** [comment]
Thanks for the updated PR. Before we move any further I'd like to make sure that we are on the same page regarding threading.

The message sequence chart you provided leaves out this important part. Please include the threads within "OSD-GDB Server," and also show how breakpoint event packets are handled.

**Comment by shivmgg** [comment]
Apologies for the delay. 
Please take a look at the updated diagram:

image

> Is the gdbserver protocol a synchronous (blocking) or asynchronous protocol?
Example: if the gdb client sends a command to read memory from the target, is the answer sent when the memory has been read, or does it return immediately (e.g. with an acknowledgement saying "command receive"), and at some later point a new message is sent with the data?

By default, whenever a packet is received, it is acknowledged with a '+' or '-' sign to establish reliability. However, we can also choose to disable the '+/-' acknowledgments by means of the `[QStartNoAckMode](http://www.it.uom.gr/teaching/gcc_manuals/onlinedocs/gdb_33.html#QStartNoAckMode)' packet.  For the current iteration, I think we should go ahead with these acknowledgment signs.

For the break command, the sequence diagram (until trap exception) should be:

image



   

> osd_hostmod provide a polling interface (osd_hostmod_event_receive()) and a asynchronous interface (register a event rx callback with osd_hostmod_new()). Depending on the answer of the remaining design both could be used.
  Questions to be answered:
      
> How are these notifications being combined with the gdb client-initiated data? I.e. breakpoint notifications be sent to a gdb client at any time, or only when no other command (see above) is being processed?



In the image, a software breakpoint (Z0) is requested at addr 0x1150. First of all, the current state of the instruction at addr 0x1150 is stored and then, a trap instruction is inserted at that address. The program continues execution until it hits the breakpoint (in this case, a trap signal). The CDM module sends a **Debug stall event packet** to the GDBserver. These event notifications should be sent immediately after the CPU stall.  GDBserver instantly clears the trap condition from the bp address when a breakpoint is hit.  
I have also included the smaller functions/threads within the GDBserver in both the diagrams. Please suggest if I should add any other information in these diagrams.

**Comment by imphil** [comment]
Your drawing shows a single thread for the osd-gdb server, which is the thread doing the TCP communication with the GDB client. You need to create this thread when osd_gdbserver_start() is called, and destroy it when osd_gdbserver_stop() is called (or however these functions will be called in the end). That should work. Please go ahead and update the code to follow this scheme.
